tests/test_template.py

```diff
@@ -0,0 +1,148 @@
+import os
+from uwtools.template import TemplateConstants, Template
+
+
+def test_substitute_string_from_dict():
+    """
+        Substitute with ${v}
+    """
+    template = '${greeting} to ${the_world}'
+    dictionary = {
+        'greeting': 'Hello',
+        'the_world': 'the world'
+    }
+    final = 'Hello to the world'
+    assert Template.substitute_structure(template,
+                                         TemplateConstants.DOLLAR_CURLY_BRACE, dictionary.get) == final
+
+
+def test_substitute_string_from_dict_paren():
+    """
+       Substitute with $(v)
+    """
+    template = '$(greeting) to $(the_world)'
+    dictionary = {
+        'greeting': 'Hello',
+        'the_world': 'the world'
+    }
+    final = 'Hello to the world'
+    assert Template.substitute_structure(template,
+                                         TemplateConstants.DOLLAR_PARENTHESES, dictionary.get) == final
+
+
+def test_assign_string_from_dict_paren():
+    """
+          Substitute with $(v) should replace with the actual object
+    """
+    template = '$(greeting)'
+    dictionary = {
+        'greeting': {
+            'a': 1,
+            'b': 2
+        }
+    }
+    assert Template.substitute_structure(template,
+                                         TemplateConstants.DOLLAR_PARENTHESES,
+                                         dictionary.get) == dictionary['greeting']
+
+
+def test_substitute_string_from_dict_double_curly():
+    """
+       Substitute with {{v}}
+    """
+    template = '{{greeting}} to {{the_world}}'
+    dictionary = {
+        'greeting': 'Hello',
+        'the_world': 'the world'
+    }
+    final = 'Hello to the world'
+    assert Template.substitute_structure(template,
+                                         TemplateConstants.DOUBLE_CURLY_BRACES,
+                                         dictionary.get) == final
+
+
+def test_substitute_string_from_dict_at_square():
+    """
+        Substitute with @[v]
+    """
+    template = '@[greeting] to @[the_world]'
+    dictionary = {
+        'greeting': 'Hello',
+        'the_world': 'the world'
+    }
+    final = 'Hello to the world'
+    assert Template.substitute_structure(template,
+                                         TemplateConstants.AT_SQUARE_BRACES,
+                                         dictionary.get) == final
+
+
+def test_substitute_string_from_dict_at_carrots():
```

Comment:
These are angle brackets. Carets (not carrots) are `^`.

Reply:
Not sure what to do with this.

Reply:
Change `carrots` to `angle_brackets` please.

Reply:
Oh yes I see they must have got the naming wrong, yes my bad I see now they simply named the test itself wrong

Reply:
done